### PR TITLE
Resolve error recovery bug relating to submode errors

### PR DESCRIPTION
**Pull request by liam923**
Resolves a bug where locating an identifier would fail when within a function application that causes a submode error.

### CODE REVIEW

File: src/ocaml/typing/typecore.ml [hunk 1]
```diff
@@ -7896,8 +7896,22 @@ and type_application env app_loc expected_mode position_and_mode
...
+             ones to fill in *)
+          record_exp_and_reraise ~exn
+            { exp_desc =
+                Texp_apply (funct, args, position_and_mode.apply_position, ap_mode);
+              exp_loc = app_loc;
+              exp_extra = [];
+              exp_type = ty_ret;
+              exp_attributes = [];
+              exp_env = env;
+            }
```
Comment (liam923) [inline src/ocaml/typing/typecore.ml:7913]: @goldfirere Please double check that these values are correct. I'm especially unsure about `exp_attributes = []`

File: tests/test-dirs/escape_region/illegal-usage.t [hunk 2]
```diff
@@ -0,0 +1,14 @@
+  $ cat >local.ml <<EOF
```
Comment (goldfirere) [inline tests/test-dirs/escape_region/illegal-usage.t:1]: Just to reduce some clutter, could the two new tests be combined into one .t file? And I would put these in the `locate` directory -- no need to create a new directory just for this case.

File: src/ocaml/typing/typecore.ml [hunk 3]
```diff
@@ -7896,8 +7896,22 @@ and type_application env app_loc expected_mode position_and_mode
       let mode_ret =
         mode_cross_left env ty_ret (alloc_as_value mode_ret)
       in
-      submode ~loc:app_loc ~env ~reason:(Application ty_ret)
-        mode_ret expected_mode;
+      let () =
+        try submode ~loc:app_loc ~env ~reason:(Application ty_ret)
+                            mode_ret expected_mode
+        with exn ->
+          (* CR lstevenson for reisenberg: I'm not sure these values are all the correct
```
Comment (goldfirere) [inline src/ocaml/typing/typecore.ml:7903]: Yep, looks good to me.

**Review by goldfirere — APPROVED** [review]

**Comment by goldfirere** [comment]
This is ready to merge. But you'll have to rebase your patches to satisfy the DCO requirement (I do not recall what that stands for). Click on the "Details" link next to the DCO failure and follow the instructions there. Once you do that and CI is green, feel free to actually merge the fix.

**Pull request merged**